Details for Turbopack loaders. (#62300)
vercel/next.js: docs/04-architecture/turbopack.mdx

}
```

## Supported features

> Turbopack in Next.js requires zero-configuration for most users and can be extended for more advanced use cases. To learn more about the currently supported features for Turbopack, view the [API Reference](/docs/architecture/turbopack).

Reviewer: ```suggestion
Turbopack in Next.js requires zero-configuration for most users and can be extended for more advanced use cases. To learn more about the currently supported features for Turbopack, view the [API Reference](/docs/app/api-reference/next-config-js/turbo).
```

Author: D'oh!

## Supported Features

Turbopack can be customized to transform different files and change how modules are resolved. It supports a subset of webpack's loader API and offers similar configuration aliasing module resolution.

> **Good to know**:
>
> Turbopack for Next.js does not require loaders nor loader configuration for
> built-in functionality, just as they aren't required for Next.js. Turbopack
> has built-in support for css and compiling modern JavaScript, so there's no
> need for `css-loader`, `postcss-loader`, or `babel-loader` if you're just
> using `@babel/preset-env`.

> If you need loader support beyond what's built in, many webpack loaders already work with Turbopack. There are currently some limitations:

Reviewer: I see, there's a list further down. Let's link to it, or maybe move the list to the top.

- Only a core subset of the webpack loader API is implemented. This is enough for some popular loaders, and we'll expand our support for this API in the future.
- Only loaders that return JavaScript code are supported. Loaders that transform files like stylesheets or images are not currently supported.

> - Options passed to webpack loaders must be plain JavaScript primitives, objects, and arrays. For example, it's not possible to pass `require()`d plugin modules as option values.

Reviewer: Does this apply to JavaScript functions in general? Or was `required()` picked for a specific purpose? 
Maybe you could say functions if it's general? 

Also, minor typo. 

```suggestion
- Options passed to webpack loaders must be plain JavaScript primitives, objects, and arrays. For example, it's not possible to pass `require()` plugin modules as option values.
```

Reviewer: Ah, I see. It's mentioned further down. Maybe you can add a heading and link to it from here. E.g. `## Importing packages as loader option`

> As of Next 13.2, configuring webpack loaders is possible for Next.js apps through an experimental option in `next.config.js`. `turbo.rules` can be set as a mapping of file globs to a list of package names or `{loader, options}` pairs:

Reviewer: We should also document this option in the `next.config.js` API reference

```js filename="next.config.js"
module.exports = {
  experimental: {
    turbo: {
      rules: {
        // Option format
        '*.md': [
          {
            loader: '@mdx-js/loader',
            options: {
              format: 'md',
            },
          },
        ],
        // Option-less format
        '*.mdx': ['@mdx-js/loader'],
      },
    },
  },
}
```

> 
> > **Good to know**:
> >
> > Note: Prior to Next.js version 13.4.4, `experimental.turbo.rules` was named
> > `experimental.turbo.loaders` and only accepted file extensions like `.mdx`
> > instead of `*.mdx`.

Reviewer: ```suggestion

> **Good to know**: Prior to Next.js version 13.4.4, `experimental.turbo.rules` was named `experimental.turbo.loaders` and only accepted file extensions like `.mdx` instead of `*.mdx`.
```

> 

Reviewer: Add heading 

```suggestion
### Importing packages as a loader option
```

```js filename="my-mdx-loader.js"
const mdxLoader = require('@mdx-js/loader')
const rehypePrism = require('@mapbox/rehype-prism')

module.exports = function (code) {
  const prevGetOptions = this.getOptions.bind(this)
  this.getOptions = function getOptions(...args) {
    return {
      ...prevGetOptions(...args),
      rehypePlugins: [rehypePrism],
    }
  }

  mdxLoader.call(this, code)
}
```

Then, configure Next.js to load the wrapper loader:

```js filename="next.config.js"
module.exports = {
  experimental: {
    turbo: {
      rules: {
        '*.mdx': ['./my-mdx-loader'],
      },
    },
  },
}
```

### Supported loaders

The following loaders have been tested to work with Turbopack's webpack loader implementation:

- [`babel-loader`](https://www.npmjs.com/package/babel-loader)
- [`@mdx-js/loader`](https://www.npmjs.com/package/@mdx-js/loader)
- [`@svgr/webpack`](https://www.npmjs.com/package/@svgr/webpack)
- [`svg-inline-loader`](https://www.npmjs.com/package/svg-inline-loader)
- [`yaml-loader`](https://www.npmjs.com/package/yaml-loader)
- [`string-replace-loader`](https://www.npmjs.com/package/string-replace-loader)

> - [`raw-loader`](https://www.npmjs.com/package/raw-loader)

Reviewer: The list of supported loaders seems very important, should we move it further up?

## Resolve aliases

Turbopack can be configured to modify module resolution through aliases, similar to webpack's [`resolve.alias`](https://webpack.js.org/configuration/resolve/#resolvealias) configuration:

```js filename="next.config.js"
module.exports = {
  experimental: {
    turbo: {
      resolveAlias: {
        underscore: 'lodash',
        mocha: { browser: 'mocha/browser-entry.js' },
      },
    },
  },
}
```

> This aliases imports of the `underscore` package to the `lodash` package so that `import underscore from 'underscore'` will load the `lodash` module instead of `underscore`.

Reviewer: Nit, for clarity. 

```suggestion
For example, this aliases the `underscore` package to the `lodash` package so that `import underscore from 'underscore'` will load the `lodash` module instead of `underscore`.
```
